// ipso/lib/loader.js
// Generated by CoffeeScript 1.6.3
var dasherize, deferred, dirname, lastInstance, lstatSync, normalize, parallel, pipeline, readdirSync, save, sep, underscore, _ref, _ref1, _ref2, _ref3;

_ref = require('also'), deferred = _ref.deferred, parallel = _ref.parallel, pipeline = _ref.pipeline;

_ref1 = require('path'), normalize = _ref1.normalize, sep = _ref1.sep, dirname = _ref1.dirname;

_ref2 = require('inflection'), underscore = _ref2.underscore, dasherize = _ref2.dasherize;

_ref3 = require('fs'), readdirSync = _ref3.readdirSync, lstatSync = _ref3.lstatSync;

save = require('./saver').save;

require('colors');

lastInstance = void 0;

module.exports._test = function() {
  return lastInstance;
};

module.exports.create = function(config) {
  var api, local;
  lastInstance = local = {
    loadModules: function(arrayOfNames, doesInstance) {
      var moduleName, promise;
      return promise = parallel((function() {
        var _i, _len, _results;
        _results = [];
        for (_i = 0, _len = arrayOfNames.length; _i < _len; _i++) {
          moduleName = arrayOfNames[_i];
          _results.push((function(moduleName) {
            return function() {
              var nestedPromise;
              return nestedPromise = pipeline([
                function() {
                  return local.loadModule(moduleName, doesInstance);
                }, function(module) {
                  return doesInstance.spectate({
                    name: moduleName,
                    tagged: false
                  }, module);
                }
              ]);
            };
          })(moduleName));
        }
        return _results;
      })());
    },
    loadModulesSync: function(arrayOfNames, doesInstance) {
      var Module, arrayOfModules, moduleName;
      return arrayOfModules = (function() {
        var _i, _len, _results;
        _results = [];
        for (_i = 0, _len = arrayOfNames.length; _i < _len; _i++) {
          moduleName = arrayOfNames[_i];
          Module = local.loadModuleSync(moduleName, doesInstance);
          _results.push(doesInstance.spectateSync({
            name: moduleName,
            tagged: false
          }, Module));
        }
        return _results;
      })();
    },
    dir: config.dir,
    modules: config.modules,
    upperCase: function(string) {
      var char, error;
      try {
        char = string[0].charCodeAt(0);
      } catch (_error) {
        error = _error;
        return false;
      }
      if (char > 64 && char < 91) {
        return true;
      }
      return false;
    },
    recurse: function(name, path, matches) {
      var fd, file, match, stat, _i, _len, _ref4, _results;
      _ref4 = readdirSync(path);
      _results = [];
      for (_i = 0, _len = _ref4.length; _i < _len; _i++) {
        fd = _ref4[_i];
        file = path + sep + fd;
        stat = lstatSync(file);
        if (stat.isDirectory()) {
          local.recurse(name, file, matches);
          continue;
        }
        if (match = fd.match(new RegExp("^(" + name + ")\.(js|coffee)$"))) {
          _results.push(matches.push(dirname(file) + sep + name));
        } else {
          _results.push(void 0);
        }
      }
      return _results;
    },
    find: function(name) {
      var matches;
      matches = [];
      try {
        local.recurse(underscore(name), local.dir + sep + 'lib', matches);
      } catch (_error) {}
      try {
        local.recurse(underscore(name), local.dir + sep + 'app', matches);
      } catch (_error) {}
      if (matches.length > 1) {
        throw new Error("ipso: found multiple matches for " + name + ", use ipso.modules");
      }
      return matches[0];
    },
    loadModule: deferred(function(action, name, does) {
      return does.get({
        query: {
          tag: name
        }
      }, function(error, spectated) {
        var dashed, mod, path;
        if (spectated != null) {
          return action.resolve(spectated.object);
        }
        if (path = ((function() {
          try {
            return local.modules[name].require;
          } catch (_error) {}
        })())) {
          if (path[0] === '.') {
            path = normalize(local.dir + sep + path);
          }
          return action.resolve(require(path));
        }
        if (!local.upperCase(name)) {
          dashed = dasherize(underscore(name));
          return action.resolve(require(dashed));
        } else {
          try {
            mod = require(name);
            return action.resolve(mod);
          } catch (_error) {}
        }
        if (path = local.find(name)) {
          return action.resolve(require(path));
        }
        console.log('ipso: ' + ("warning: missing module " + name).yellow);
        return action.resolve({
          $ipso: {
            PENDING: true,
            module: name
          },
          $save: function(template) {
            if (template == null) {
              template = 'default';
            }
            return save(template, name, does);
          }
        });
      });
    }),
    loadModuleSync: function(name, does) {
      var Module, dashed, path;
      if (Module = ((function() {
        try {
          return does.getSync({
            query: {
              tag: name
            }
          });
        } catch (_error) {}
      })())) {
        return Module;
      }
      if (path = ((function() {
        try {
          return local.modules[name].require;
        } catch (_error) {}
      })())) {
        console.log({
          path: path
        });
        if (path[0] === '.') {
          path = normalize(local.dir + sep + path);
        }
        return require(path);
      }
      if (!local.upperCase(name)) {
        dashed = dasherize(underscore(name));
        return require(dashed);
      } else {
        try {
          return require(name);
        } catch (_error) {}
      }
      if (path = local.find(name)) {
        return require(path);
      }
      console.log('ipso: ' + ("warning: missing module " + name).yellow);
      return {
        $ipso: {
          PENDING: true,
          module: name
        },
        $save: function(template) {
          if (template == null) {
            template = 'default';
          }
          return save(template, name, does);
        }
      };
    }
  };
  return api = {
    loadModules: local.loadModules,
    loadModulesSync: local.loadModulesSync
  };
};
